### PR TITLE
Hadoop comptible exception mkdir

### PR DESCRIPTION
**Pull request by minmingzhu**
**The mkdirs function needs to meet the following conditions:**
1. Create a directory and all its parents
2. if exists(FS,P) and not isDir(FS,p) ,needs to raise [ParentNotDirectoryException, FileAlreadyExistsException, IOException]  - we use DAOS error code by native code to translate 
 and throw exception. 
3. The local FileSystem does not raise an exception  if mkdirs(p) is invoked on a path that exists and is a file. Instead the operation returns false .  - The native code didn't support this , when  invoked on a path that exists and is a file , it return true . Therefore , we support this operating in Java Wrapper.  

**About unit testing :**
The test consists of 2 parts:
1. TestDaosContractMkdir : we need to setup pool uuid ,  container uuid and svc in resources/auth-keys.html.

```
 <property> 
        <name>fs.daos.pool.uuid</name>   
        <value>e03ec7b9-4d12-4869-af35-bb634e79848a</value>
    </property>
    <property>   
        <name>fs.daos.container.uuid</name>   
        <value>93c58d34-373a-4b42-9ca7-c465efe0de23</value>
    </property>
    <property>
        <name>fs.daos.pool.svc</name>
        <value>0</value>
    </property>
```
2. DaosFileSystemContractIT, in this test method we only need to focus on the test cases of mkdirs.

**_This is Hadoop Comptible Exception doc about mkdirs ._** 

**boolean mkdirs(Path p, FsPermission permission )**
Create a directory and all its parents

**Preconditions**
 ```
if exists(FS, p) and not isDir(FS, p) :
     raise [ParentNotDirectoryException, FileAlreadyExistsException, IOException]
```
**Postconditions**
```
FS' where FS'.Directories' = FS.Directories + [p] + ancestors(FS, p)
result = True
```
The condition exclusivity requirement of a FileSystem’s directories, files and symbolic links must hold.

The probe for the existence and type of a path and directory creation MUST be atomic. The combined operation, including mkdirs(parent(F)) MAY be atomic.

The return value is always true—even if a new directory is not created (this is defined in HDFS).

**Implementation Notes: Local FileSystem**
The local FileSystem does not raise an exception if mkdirs(p) is invoked on a path that exists and is a file. Instead the operation returns false.

```
if isFile(FS, p):
   FS' = FS
   result = False
```

### CODE REVIEW

File: src/client/java/daos-java/pom.xml [hunk 1]
```diff
@@ -45,6 +45,12 @@
       <artifactId>junit</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>2.7.6</version>
+      <scope>compile</scope>
+    </dependency>
```
Comment (zjf2012) [inline src/client/java/daos-java/pom.xml:53]: no hadoop reference in daos-java module which is not just for hadoop daos

File: src/client/java/hadoop-daos/src/test/resources/auth-keys.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        <value>${test.fs.daos.name}</value>
+    </property>
+
+    <property>
+        <name>fs.daos.impl</name>
+        <value>io.daos.fs.hadoop.DaosFileSystem</value>
+    </property>
+    <property> 
+        <name>fs.daos.pool.uuid</name>   
+        <value>e03ec7b9-4d12-4869-af35-bb634e79848a</value>
```
Comment (zjf2012) [inline src/client/java/hadoop-daos/src/test/resources/auth-keys.xml:20]: the uuids need to be able to be passed from command line in integration test. please check existing way of passing the uuids.

File: src/client/java/daos-java/pom.xml (hunk 1 above)
Comment (minmingzhu) [reply src/client/java/daos-java/pom.xml:53]: mkdirs method needs to throw ParentNotDirectoryException , when parent directory did not exist . This exception belong  org.apache.hadoop.fs

File: src/client/java/hadoop-daos/src/test/resources/auth-keys.xml (hunk 2 above)
Comment (minmingzhu) [reply src/client/java/hadoop-daos/src/test/resources/auth-keys.xml]: it done

File: src/client/java/daos-java/pom.xml [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
   <properties>
     <native.build.path>${project.basedir}/build</native.build.path>
-    <daos.install.path>/usr/local/daos</daos.install.path>
+    <daos.install.path>/home/damon/cluster/daos_200408/install</daos.install.path>
```
Comment (zjf2012) [inline src/client/java/daos-java/pom.xml:19]: please restore it to /usr/local/daos

File: src/client/java/daos-java/pom.xml (hunk 1 above)
Comment (zjf2012) [reply src/client/java/daos-java/pom.xml:53]: still there?

File: src/client/java/pom.xml [hunk 4]
```diff
@@ -33,7 +33,7 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <dependency.locations.enabled>false</dependency.locations.enabled>
     <native.build.path>${project.basedir}/build</native.build.path>
-    <daos.install.path>/usr/local/daos</daos.install.path>
+    <daos.install.path>/home/damon/cluster/daos_200408/install</daos.install.path>
```
Comment (zjf2012) [inline src/client/java/pom.xml:36]: please restore it to /usr/local/daos

File: src/client/java/daos-java/src/main/java/io/daos/dfs/DaosUtils.java [hunk 5]
```diff
@@ -107,4 +122,33 @@ public static long toMilliSeconds(StatAttributes.TimeSpec timeSpec) {
...
+        }else{
+          ioe = new NotDirectoryException(daosIOException.getMessage());
+        }
+        break;
+      default :
+        ioe = daosIOException;
+        break;
+    }
+    return ioe;
+  }
```
Comment (zjf2012) [inline src/client/java/daos-java/src/main/java/io/daos/dfs/DaosUtils.java]: come back?

File: src/client/java/hadoop-daos/pom.xml [hunk 6]
```diff
@@ -17,7 +17,7 @@
   <properties>
     <hadoop.version>2.7.6</hadoop.version>
     <native.build.path>${project.basedir}/build</native.build.path>
-    <daos.install.path>/usr/local/daos</daos.install.path>
+    <daos.install.path>/home/damon/cluster/daos_200408/install</daos.install.path>
```
Comment (zjf2012) [inline src/client/java/hadoop-daos/pom.xml]: please restore it to /usr/local/daos

**Pull request merged**